contributor-docs/components/dependencies.md

```diff
@@ -3,5 +3,187 @@ parent: Firebase Components
+to delay it until it's actually needed.
+
+It is also benefitial to use a `Provider` in the context of [Play's dynamic feature delivery](https://developer.android.com/guide/playcore/feature-delivery).
+See [Dynamic Module Support]({{ site.baseurl }}{% link components/dynamic_modules.md %}) for more details.
+
+## Required dependencies
+
+This type of dependency informs the `ComponentRuntime` that a given `Component` cannot function without a dependency.
+When the dependency is missing during initialization, `ComponentRuntime` will throw a `DependencyMissingDependency`.
+This type of dependency is useful for built-in components that are always present like `Context`, `FirebaseApp`,
+`FirebaseOptions`, [Executors]({{ site.baseurl }}{% link components/executors.md %}).
+
+To declare a required dependency use one of the following in your `ComponentRegistrar`:
+
+```java
+  // Required directly injected dependency
+  .add(Dependency.required(MyDep.class))
+  // Required lazily injected dependency
+  .add(Dependency.requiredProvider(MyOtherDep.class))
+  .factory( c -> new MyComponent(c.get(MyDep.class), c.getProvider(MyOtherDep.class)))
+  .build();
+```
+
+## Optional Dependencies
+
+This type of dependencies is useful when your `Component` can operate normally when the dependency is not
+available, but can have enhanced functionality when present. e.g. `Firestore` can work without `Auth` but
+provides secure database access when `Auth` is present.
+
+To declare a required dependency use the following in your `ComponentRegistrar`:
+
+```java
+  .add(Dependency.optionalProvider(MyDep.class))
+  .factory(c -> new MyComponent(c.getProvider(MyDep.class)))
+  .build();
+```
+
+The provider will return `null` if the dependency is not present in the app.
+
+{: .warning }
+When the app uses [Play's dynamic feature delivery](https://developer.android.com/guide/playcore/feature-delivery),
+`provider.get()` can return your dependency when it becomes available, so be prepared for that.
+And don't store references to the result of `provider.get()` calls.
+
+See [Dynamic Module Support]({{ site.baseurl }}{% link components/dynamic_modules.md %}) for details
+
+{: .warning }
+See Deferred dependencies if you your dependency has a callback based apis
+
+## Deferred Dependencies
+
+Useful for optional dependencies which have a listener-style API, i.e. the dependent component registers a
+listener with the dependency and never calls it again(instead the dependency will call the registered listener).
+A good example is `Firestore`'s use of `Auth` where `Firestore` registers a token change listener to get
+notified when a new token is available. The problem is that when `Firestore` initializes, `Auth` may not be
+present in the app, and is instead part of a dynamic module that can be loaded at runtime on demand.
+
+To solve this problem, Components have a notion of a `Deferred` dependency. A deferred is defined as follows:
```

Comment:
```suggestion
To solve this problem, Components have a notion of the of a `Deferred` dependency. A deferred is defined as follows:
```

Reply:
this does not look grammatically correct

Reply:
Yes, absolutely. Please ignore